nidirect_breadcrumbs/src/NodeThemesBreadcrumb.php

```diff
@@ -133,6 +133,26 @@ public function build(RouteMatchInterface $route_match) {
       }
     }
 
+    // Determine if node is part of a book and add link(s) to its book parent(s).
+    if (!empty($node->book)) {
+
+      // Determine depth of node in the book.
+      $book_depth = $node->book['depth'];
+
+      // Create links to parent nodes in the depths above - if published.
+      $i = 1;
+      while ($i < $book_depth) {
+        $p = 'p' . $i++;
+        $nid_published = \Drupal::entityQuery('node')
```

Comment:
As a rule of thumb, when you have a custom class using direct service calls (`\Drupal::serviceName` or `Drupal::service('service_id')`) you will want to pass these in using dependency injection. That's not normally as hard as it sounds... usually a constructor, a static create function and some internal variables to represent the service objects.

In this instance, I'll propose progressive optimisation. So merge this (because it works and is safe) but make a note or JIRA ticket to swing back and introduce injected services to this class.